src/traction_modem/Link.hxx

```diff
@@ -0,0 +1,404 @@
+class Link : private Atomic
+{
+public:
+    /// Constructor
+    /// @param tx_iface reference to the transmit interface
+    /// @param rx_iface reference to the receive interface
+    Link(TxInterface *tx_iface, RxInterface *rx_iface)
+        : txIface_(tx_iface)
+        , rxIface_(rx_iface)
+        , state_(State::STOP)
+    {
+    }
+
+    /// Check on the link status
+    /// @return true if link is up, else link is down.
+    bool is_link_up()
+    {
+        return state_ == State::UP;
+    }
+
+    /// Bind an interface to the TX and RX interfaces to start transmitting and
+    /// receiving on.
+    /// @param fd interface to transmit and receive messages on
+    void start(int fd)
+    {
+        {
+            AtomicHolder h(this);
+            if (state_ != State::STOP)
+            {
+                // Already started.
+                return;
+            }
+            state_ = State::DOWN;
+        }        
+        txIface_->start(fd);
+        rxIface_->start(fd);
+        for (auto it = linkIfaces_.begin(); it != linkIfaces_.end(); ++it)
+        {
+            (*it)->link_start();
+        }
+    }
+
+    /// Register for updates on link status.
+    /// @param interface object to register
+    void register_link_status(LinkStatusInterface *interface)
+    {
+        linkIfaces_.push_back(interface);
+    }
+
+    /// Register for updates on link status.
+    /// @param interface object to register
+    void unregister_link_status(LinkStatusInterface *interface)
+    {
+        for (auto it = linkIfaces_.begin(); it != linkIfaces_.end(); ++it)
+        {
+            if (interface == *it)
+            {
+                linkIfaces_.erase(it);
+                break;
+            }
+        }
+    }
+
+    /// Get a reference to the link's transmit interface.
+    TxInterface *get_tx_iface()
+    {
+        return txIface_;
+    }
+
+    /// Get a reference to the links receive interface.
+    RxInterface *get_rx_iface()
+    {
+        return rxIface_;
+    }
+
+private:
+    /// Link state.
+    enum class State
+    {
+        STOP, ///< link not yet started
+        DOWN, ///< link down
+        UP, ///< link up
+    };
+
+    /// Called when link transitions to "up" state.
+    void link_up()
+    {
+        state_ = State::UP;
+        for (auto it = linkIfaces_.begin(); it != linkIfaces_.end(); ++it)
+        {
+            (*it)->link_up();
+        }
+    }
+
+    /// Called when link transitions to "down" state.
+    void link_down()
+    {
+        state_ = State::DOWN;
+        for (auto it = linkIfaces_.begin(); it != linkIfaces_.end(); ++it)
+        {
+            (*it)->link_down();
+        }
+    }
+
+    TxInterface *txIface_; ///< reference to the transmit interface interface
+    RxInterface *rxIface_; ///< reference to the receive interface
+    State state_; ///< link state
+
+    /// List of interfaces that have registered for link updates.
+    std::vector<LinkStatusInterface*> linkIfaces_;
```

Comment:
Consider some better name for this variable, like linkCallbacks_.

In general, calling a variable Iface_ when that variable is not a network interface makes the code hard to read. Specifically, here "iface" seems to just mean an abstract class. However, the variable is never an abstract class. It is a pointer to an actual implementation of that abstract class and under no circumstances is the variable an interface.